modules/foreman_subnet.py

```diff
@@ -0,0 +1,315 @@
+    'dns_proxy': 'dns',
+    'remote_execution_proxies': 'remote_execution_proxy',
+    'vlanid': 'vlanid',
+    'mtu': 'mtu',
+    'organizations': 'organization',
+    'locations': 'location',
+}
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            server_url=dict(required=True),
+            username=dict(required=True),
+            password=dict(required=True, no_log=True),
+            verify_ssl=dict(type='bool', default=True),
+            name=dict(required=True),
+            network_type=dict(choices=['IPv4', 'IPv6'], default='IPv4'),
+            dns_primary=dict(),
+            dns_secondary=dict(),
+            domains=dict(type='list'),
+            gateway=dict(),
+            network=dict(required=True),
+            cidr=dict(type='int'),
+            mask=dict(),
+            from_ip=dict(),
+            to_ip=dict(),
+            boot_mode=dict(choices=['DHCP', 'Static'], default='DHCP'),
+            ipam=dict(choices=['DHCP', 'Internal DB'], default='DHCP'),
+            dhcp_proxy=dict(),
+            tftp_proxy=dict(),
+            discovery_proxy=dict(),
+            dns_proxy=dict(),
+            remote_execution_proxies=dict(type='list'),
+            vlanid=dict(type='int'),
+            mtu=dict(type='int'),
+            locations=dict(type='list'),
+            organizations=dict(type='list'),
+            state=dict(choices=['present', 'absent'], default='present'),
+        ),
+        required_one_of=[['cidr', 'mask']],
+        supports_check_mode=True,
+    )
+
+    if has_import_error:
+        module.fail_json(msg=import_error_msg)
+
+    entity_dict = dict(
+        [(k, v) for (k, v) in module.params.items() if v is not None])
+
+    server_url = entity_dict.pop('server_url')
+    username = entity_dict.pop('username')
+    password = entity_dict.pop('password')
+    verify_ssl = entity_dict.pop('verify_ssl')
+    state = entity_dict.pop('state')
+
+    try:
+        create_server(server_url, (username, password), verify_ssl)
+    except Exception as e:
+        module.fail_json(msg="Failed to connect to Foreman server: %s " % e)
+
+    ping_server(module)
+    try:
+        # Try to find the Subnet to work on
+        entity = find_subnet(module, name=entity_dict['name'], failsafe=True)
+    except Exception as e:
+        module.fail_json(msg='Failed to find entity: %s ' % e)
+
+    if 'mask' in entity_dict and 'cidr' not in entity_dict:
+        entity_dict['cidr'] = IPNetwork('%s/%s' % (entity_dict['network'], entity_dict['mask'])).prefixlen
+    elif 'mask' not in entity_dict and 'cidr' in entity_dict:
+        entity_dict['mask'] = IPNetwork('%s/%s' % (entity_dict['network'], entity_dict['cidr'])).netmask
+
+    if 'domains' in entity_dict:
+        entity_dict['domains'] = find_domains(module, entity_dict['domains'])
+
+    if 'dhcp_proxy' in entity_dict:
+        entity_dict['dhcp_proxy'] = find_smart_proxy(module, entity_dict['dhcp_proxy'])
```

Comment:
Mind re-writing these ` find_smart_proxy` with a for loop, maybe something like will work:

```python
for feature in ('tftp_proxy',  'dhcp_proxy' ,'ect...'):
   entity_dict[feature] = find_smart_proxy(module, entity_dict[feature])